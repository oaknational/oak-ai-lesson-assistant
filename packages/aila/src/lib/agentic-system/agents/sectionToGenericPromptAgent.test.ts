import { z } from "zod";

import type { LooseLessonPlan } from "../../../protocol/schema";
import type {
  AgenticRagLessonPlanResult,
  AilaExecutionContext,
  ChatMessage,
  SectionPromptAgentProps,
} from "../types";
import type { VoiceId } from "./sectionAgents/shared/voices";
import { sectionToGenericPromptAgent } from "./sectionToGenericPromptAgent";

// Mock schema for testing
const mockSchema = z.object({
  content: z.string(),
  items: z.array(z.string()).optional(),
});

type MockSchemaType = z.infer<typeof mockSchema>;

describe("sectionToGenericPromptAgent", () => {
  const mockMessages: ChatMessage[] = [
    {
      id: "msg-1",
      role: "user",
      content:
        "Can you help me create a lesson about photosynthesis for Year 7 students?",
    },
    {
      id: "msg-2",
      role: "assistant",
      content:
        "I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    },
    {
      id: "msg-3",
      role: "user",
      content:
        "Please make sure to include the key learning points about chlorophyll and glucose production.",
    },
  ];

  const mockDocument: LooseLessonPlan = {
    title: "Introduction to Photosynthesis",
    keyStage: "ks3",
    subject: "science",
    topic: "Plants and ecosystems",
    learningOutcome: "Students will understand the process of photosynthesis",
  };

  const mockRelevantLessons: AgenticRagLessonPlanResult[] = [];

  const mockExecutionContext: AilaExecutionContext = {
    persistedState: {
      messages: mockMessages,
      initialDocument: mockDocument,
      relevantLessons: mockRelevantLessons,
    },
    runtime: {
      plannerAgent: jest.fn(),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      sectionAgents: {} as any, // Mock - not used in this test
      messageToUserAgent: jest.fn(),
      fetchRelevantLessons: jest.fn(),
      config: {
        mathsQuizEnabled: false,
      },
    },
    currentTurn: {
      document: mockDocument,
      plannerOutput: null,
      errors: [],
      stepsExecuted: [],
      relevantLessons: [],
      relevantLessonsFetched: false,
    },
    callbacks: {
      onPlannerComplete: jest.fn(),
      onSectionComplete: jest.fn(),
      onTurnComplete: jest.fn(),
    },
  };

  const mockContentToString = (content: MockSchemaType): string => {
    if (!content) return "";
    return `${content.content}${content.items ? ` - ${content.items.join(", ")}` : ""}`;
  };

  const baseProps: SectionPromptAgentProps<MockSchemaType> = {
    responseSchema: mockSchema,
    messages: mockMessages,
    instructions:
      "Generate key learning points for a science lesson about photosynthesis.",
    currentValue: undefined,
    exemplarContent: undefined,
    basedOnContent: undefined,
    contentToString: mockContentToString,
    ctx: mockExecutionContext,
  };

  describe("input message generation", () => {
    it("should generate consistent input messages with minimal props", () => {
      const agent = sectionToGenericPromptAgent(baseProps);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle current value", () => {
      const currentValue: MockSchemaType = {
        content: "Photosynthesis is the process plants use to make food",
        items: ["chlorophyll", "sunlight", "water"],
      };

      const propsWithCurrentValue: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        currentValue,
      };

      const agent = sectionToGenericPromptAgent(propsWithCurrentValue);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle exemplar content", () => {
      const exemplarContent: MockSchemaType[] = [
        {
          content: "Respiration is the process animals use to release energy",
          items: ["oxygen", "glucose", "carbon dioxide"],
        },
        {
          content: "Digestion breaks down food into useful nutrients",
          items: ["enzymes", "nutrients", "absorption"],
        },
      ];

      const propsWithExemplar: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        exemplarContent,
      };

      const agent = sectionToGenericPromptAgent(propsWithExemplar);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle based on content", () => {
      const basedOnContent: MockSchemaType = {
        content: "Plant structure and function basics",
        items: ["leaves", "roots", "stem"],
      };

      const propsWithBasedOn: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        basedOnContent,
      };

      const agent = sectionToGenericPromptAgent(propsWithBasedOn);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle default voice", () => {
      const propsWithVoice: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        defaultVoice: "TEACHER_TO_PUPIL_WRITTEN" as VoiceId,
      };

      const agent = sectionToGenericPromptAgent(propsWithVoice);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle multiple voices", () => {
      const propsWithVoices: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        defaultVoice: "AILA_TO_TEACHER" as VoiceId,
        voices: ["AILA_TO_TEACHER", "TEACHER_TO_PUPIL_WRITTEN"] as VoiceId[],
      };

      const agent = sectionToGenericPromptAgent(propsWithVoices);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle extra input from context", () => {
      const extraInputFromCtx = (_ctx: AilaExecutionContext) => [
        {
          role: "developer" as const,
          content:
            "Additional context: This is a Year 7 science lesson focused on biological processes.",
        },
        {
          role: "user" as const,
          content:
            "Remember to keep explanations age-appropriate for 11-12 year-old students.",
        },
      ];

      const propsWithExtra: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        extraInputFromCtx,
      };

      const agent = sectionToGenericPromptAgent(propsWithExtra);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle all options combined", () => {
      const currentValue: MockSchemaType = {
        content: "Basic photosynthesis overview",
        items: ["plants", "energy", "sunlight"],
      };

      const exemplarContent: MockSchemaType[] = [
        {
          content: "Cellular respiration example",
          items: ["mitochondria", "ATP", "glucose"],
        },
      ];

      const basedOnContent: MockSchemaType = {
        content: "Foundation knowledge about plant biology",
        items: ["cell structure", "organelles", "life processes"],
      };

      const extraInputFromCtx = (_ctx: AilaExecutionContext) => [
        {
          role: "developer" as const,
          content: "Context: Advanced Year 7 class with strong prior knowledge",
        },
      ];

      const complexProps: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        currentValue,
        exemplarContent,
        basedOnContent,
        extraInputFromCtx,
        defaultVoice: "EXPERT_TEACHER" as VoiceId,
        voices: ["EXPERT_TEACHER", "TEACHER_TO_PUPIL_WRITTEN"] as VoiceId[],
      };

      const agent = sectionToGenericPromptAgent(complexProps);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle empty messages array", () => {
      const propsWithNoMessages: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        messages: [],
      };

      const agent = sectionToGenericPromptAgent(propsWithNoMessages);

      expect(agent.input).toMatchSnapshot();
    });

    it("should handle different instruction types", () => {
      const detailedInstructions = `
Generate comprehensive key learning points for a photosynthesis lesson.

Requirements:
- Include 3-5 main points
- Focus on the process, inputs, and outputs
- Make connections to everyday life
- Consider common misconceptions

The content should be suitable for Key Stage 3 students.
      `.trim();

      const propsWithDetailedInstructions: SectionPromptAgentProps<MockSchemaType> =
        {
          ...baseProps,
          instructions: detailedInstructions,
        };

      const agent = sectionToGenericPromptAgent(propsWithDetailedInstructions);

      expect(agent.input).toMatchSnapshot();
    });
  });

  describe("schema validation", () => {
    it("should have the correct response schema", () => {
      const agent = sectionToGenericPromptAgent(baseProps);

      expect(agent.responseSchema).toBe(mockSchema);
      expect(typeof agent.responseSchema.parse).toBe("function");
    });
  });

  describe("voice handling", () => {
    it("should add default voice to voices array if not present", () => {
      const propsWithVoiceNotInArray: SectionPromptAgentProps<MockSchemaType> =
        {
          ...baseProps,
          defaultVoice: "AILA_TO_TEACHER" as VoiceId,
          voices: ["TEACHER_TO_PUPIL_WRITTEN"] as VoiceId[],
        };

      const agent = sectionToGenericPromptAgent(propsWithVoiceNotInArray);

      expect(agent.input).toMatchSnapshot();
    });

    it("should not duplicate default voice if already in voices array", () => {
      const propsWithVoiceInArray: SectionPromptAgentProps<MockSchemaType> = {
        ...baseProps,
        defaultVoice: "AILA_TO_TEACHER" as VoiceId,
        voices: ["AILA_TO_TEACHER", "TEACHER_TO_PUPIL_WRITTEN"] as VoiceId[],
      };

      const agent = sectionToGenericPromptAgent(propsWithVoiceInArray);

      expect(agent.input).toMatchSnapshot();
    });
  });
});
