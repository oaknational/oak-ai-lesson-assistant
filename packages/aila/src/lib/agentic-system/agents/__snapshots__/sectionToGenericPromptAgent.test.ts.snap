// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sectionToGenericPromptAgent input message generation should generate consistent input messages with minimal props 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle all options combined 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### EXPERT_TEACHER
Speaker: Expert teacher
Audience: User
Use this voice to explain (from your experience) key knowledge, common mistakes and misconceptions that pupils at this age might have or need, pedagogical insights and advice.

### TEACHER_TO_PUPIL_WRITTEN
Speaker: Teacher
Audience: Pupils (written)
Use this voice to write the text that will appear on slides or lesson resources. Formal, concise, and instructional. No chatty tone.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use EXPERT_TEACHER.",
    "role": "developer",
  },
  {
    "content": "CURRENT SECTION VALUE

This is the current value of the section you are creating. Use it as a starting point.

Basic photosynthesis overview - plants, energy, sunlight",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

<example-1>Cellular respiration example - mitochondria, ATP, glucose</example-1>",
    "role": "developer",
  },
  {
    "content": "BASED ON CONTENT

Only diverge from this content if necessary to make sense or if the user has specified otherwise. If there's no way to do this without contradiction then return an error, with justification.

Foundation knowledge about plant biology - cell structure, organelles, life processes",
    "role": "developer",
  },
  {
    "content": "Context: Advanced Year 7 class with strong prior knowledge",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle based on content 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "BASED ON CONTENT

Only diverge from this content if necessary to make sense or if the user has specified otherwise. If there's no way to do this without contradiction then return an error, with justification.

Plant structure and function basics - leaves, roots, stem",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle current value 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "CURRENT SECTION VALUE

This is the current value of the section you are creating. Use it as a starting point.

Photosynthesis is the process plants use to make food - chlorophyll, sunlight, water",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle default voice 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### TEACHER_TO_PUPIL_WRITTEN
Speaker: Teacher
Audience: Pupils (written)
Use this voice to write the text that will appear on slides or lesson resources. Formal, concise, and instructional. No chatty tone.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use TEACHER_TO_PUPIL_WRITTEN.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle different instruction types 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate comprehensive key learning points for a photosynthesis lesson.

Requirements:
- Include 3-5 main points
- Focus on the process, inputs, and outputs
- Make connections to everyday life
- Consider common misconceptions

The content should be suitable for Key Stage 3 students.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle empty messages array 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

There have been no previous messages. See user's message below, which is the start of the interaction.

",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle exemplar content 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

<example-1>Respiration is the process animals use to release energy - oxygen, glucose, carbon dioxide</example-1>
<example-2>Digestion breaks down food into useful nutrients - enzymes, nutrients, absorption</example-2>",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle extra input from context 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "Additional context: This is a Year 7 science lesson focused on biological processes.",
    "role": "developer",
  },
  {
    "content": "Remember to keep explanations age-appropriate for 11-12 year-old students.",
    "role": "user",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent input message generation should handle multiple voices 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.

### TEACHER_TO_PUPIL_WRITTEN
Speaker: Teacher
Audience: Pupils (written)
Use this voice to write the text that will appear on slides or lesson resources. Formal, concise, and instructional. No chatty tone.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent voice handling should add default voice to voices array if not present 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.

### TEACHER_TO_PUPIL_WRITTEN
Speaker: Teacher
Audience: Pupils (written)
Use this voice to write the text that will appear on slides or lesson resources. Formal, concise, and instructional. No chatty tone.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;

exports[`sectionToGenericPromptAgent voice handling should not duplicate default voice if already in voices array 1`] = `
[
  {
    "content": "# Identity

You are an agent as part of Aila, an AI-powered lesson planning assistant designed for use by UK teachers.
In collaboration with the user, Aila creates a lesson plan.
You will be given a specific task to complete and will be provided with the current state of the lesson plan and recent messages between the user and the assistant, which might give insight into the user's needs and how to best support them.

## Markdown
Do not use markdown formatting unless unless specified for a specific section.

## Language
Use British English spelling and vocabulary (e.g. colour not color, centre not centre, rubbish not trash) unless the user sets a different primary language. This reflects our UK teacher audience.
",
    "role": "developer",
  },
  {
    "content": "Generate key learning points for a science lesson about photosynthesis.",
    "role": "developer",
  },
  {
    "content": "## Voice Definitions

The following voice definitions are available for this task:

### AILA_TO_TEACHER
Speaker: Aila
Audience: User
Supportive expert guiding the teacher. Polite and clear. You can ask for clarification from the teacher if required.

### TEACHER_TO_PUPIL_WRITTEN
Speaker: Teacher
Audience: Pupils (written)
Use this voice to write the text that will appear on slides or lesson resources. Formal, concise, and instructional. No chatty tone.",
    "role": "developer",
  },
  {
    "content": "## Voice guidance

Unless otherwise specified, use AILA_TO_TEACHER.",
    "role": "developer",
  },
  {
    "content": "EXEMPLAR CONTENT

This is exemplar content from similar Oak Academy lessons, it should be used as a guide for the section you are creating.

",
    "role": "developer",
  },
  {
    "content": "MESSAGE HISTORY

These are the previous messages in the current conversation with the user.

- user: Can you help me create a lesson about photosynthesis for Year 7 students?
- assistant: I'd be happy to help you create a lesson about photosynthesis for Year 7 students.",
    "role": "developer",
  },
  {
    "content": "USER's MESSAGE

This is the most recent message from the user that we've been attempting to process.

Please make sure to include the key learning points about chlorophyll and glucose production.",
    "role": "user",
  },
]
`;
